### ISSUE TITLE
`useLastNotificationResponse` returns the interacted notificaiton twice when app was killed

### ISSUE THREAD

**Issue by itsramiel** [expo/expo#30649]
### Minimal reproducible example

https://github.com/itsramiel/double-notification

### What platform(s) does this occur on?

iOS

### Where did you reproduce the issue?

in a development build, in Expo Go, in a standalone app

### Summary

The `useLastNotificationResponse` hook updates with the same notification object twice when clicking on a notification object and app is killed. This affects apps that depend on this notification and do something according to it.

For example I have a useEffect with a dependency array of the notification object received from this hook. But since the hook rerenders/sets state for the same notification object I do the action inside the useEffect twice.

I checked the source code and it seem like the hook sets the notification once in a `useLayoutEffect` and another in a `useEffect` which seems to be causing the issue:

https://github.com/expo/expo/blob/42ffc110ea1ba526703dab3ea07b812254c111ef/packages/expo-notifications/src/useLastNotificationResponse.ts#L50-L67

I added a console log in expo's source code and yes both return the same notification object and both set the state cause 2 rerenders:
<img width="1725" alt="image" src="https://github.com/user-attachments/assets/a788ffc6-8306-4b5c-b5d5-f3d746b3c64b">

For my own project I believe I will go with the lower level `getLastNotificationResponseAsync`, and I believe the hook should just be using one effect or?

### Environment

```
  expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.3.1
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 20.11.1 - ~/.nvm/versions/node/v20.11.1/bin/node
      Yarn: 1.22.22 - ~/.nvm/versions/node/v20.11.1/bin/yarn
      npm: 10.2.4 - ~/.nvm/versions/node/v20.11.1/bin/npm
      Watchman: 2024.06.10.00 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.14.3 - /Users/itsramiel/.rbenv/shims/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.5, iOS 17.5, macOS 14.5, tvOS 17.5, visionOS 1.2, watchOS 10.5
    IDEs:
      Android Studio: 2023.1 AI-231.9392.1.2311.11330709
      Xcode: 15.4/15F31d - /usr/bin/xcodebuild
    npmPackages:
      expo: ~51.0.22 => 51.0.22
      react: 18.2.0 => 18.2.0
      react-native: 0.74.3 => 0.74.3
    npmGlobalPackages:
      eas-cli: 10.2.0
    Expo Workflow: managed
    ```

### Expo Doctor Diagnostics

```
~/Des/A/T/e/Notifications main !2 ❯ npx expo-doctor@latest                                   08:33:23 PM
✔ Check Expo config for common issues
✔ Check package.json for common issues
✔ Check native tooling versions
✔ Check dependencies for packages that should not be installed directly
✔ Check for common project setup issues
✔ Check for issues with metro config
✔ Check npm/ yarn versions
✔ Check Expo config (app.json/ app.config.js) schema
✔ Check that packages match versions required by installed Expo SDK
✔ Check for legacy global CLI installed locally
✔ Check that native modules do not use incompatible support packages
✔ Check that native modules use compatible support package versions for installed Expo SDK

Didn't find any issues with the project!
```

**Comment by douglowder** [comment]
Good catch -- will fix that!

**Comment by itsramiel** [comment]
Okay so according to  expo docs `NotificationReceivedListener` is triggered when a user interacts with a notification while app is in background or foreground
<img width="1085" alt="image" src="https://github.com/user-attachments/assets/ff826853-1623-4a58-8ce0-7576b83cc080">

However on iOS it is also triggered when app is killed, causing the `useLastNotificationResponse` to return the same notification twice once from `getLastNotificationResponseAsync` as expected another from `NotificationResponseReceivedListener` which I believe is not expected and is the bug.

**Comment by itsramiel** [comment]
Okay wow, so I am finding some really interesting/buggy findings.

## Testing iOS only since that is where my issue is

If a user taps on a notification while app is closed `getLastNotificationResponseAsync` returns nothing when used by itself in such an app:

```
export default function App() {
  const [expoPushToken, setExpoPushToken] = useState("");

  useEffect(() => {
    Notifications.getLastNotificationResponseAsync().then(
      (lastNotification) => {
        if (lastNotification) {
          Alert.alert("Notification", JSON.stringify(lastNotification));
        }
      },
    );
  }, []);

  useEffect(() => {
    registerForPushNotificationsAsync()
      .then((token) => setExpoPushToken(token ?? ""))
      .catch((error: any) => setExpoPushToken(`${error}`));
  }, []);

  return (
    <View
      style={{ flex: 1, alignItems: "center", justifyContent: "space-around" }}
    >
      <Text>Your Expo push token: {expoPushToken}</Text>
    </View>
  );
}
```

If a `NotificationResponseReceivedListener` is setup and `getLastNotificationResponseAsync` is called then `getLastNotificationResponseAsync` returns the last notification and the listener is fired for a notification that is interacted with when an app is killed:
```
export default function App() {
  const [expoPushToken, setExpoPushToken] = useState("");

  useEffect(() => {
    Notifications.getLastNotificationResponseAsync().then(
      (lastNotification) => {
        if (lastNotification) {
          Alert.alert("Notification", JSON.stringify(lastNotification));
        }
      },
    );

    const sub = Notifications.addNotificationResponseReceivedListener(
      (response) => {
        Alert.alert("Notification", JSON.stringify(response));
      },
    );

    return sub.remove;
  }, []);

  useEffect(() => {
    registerForPushNotificationsAsync()
      .then((token) => setExpoPushToken(token ?? ""))
      .catch((error: any) => setExpoPushToken(`${error}`));
  }, []);

  return (
    <View
      style={{ flex: 1, alignItems: "center", justifyContent: "space-around" }}
    >
      <Text>Your Expo push token: {expoPushToken}</Text>
    </View>
  );
}
```

It is as if `getLastNotificationResponseAsync` depends on the listener being set.

**Comment by douglowder** [comment]
I'll have a look at this example, to see why this behavior occurs. I've been thinking about this issue a bit more.

In general, I think it's probably expected that `getLastNotificationResponseAsync()` and `addNotificationResponseReceivedListener()` might have the same response, especially since `getLastNotificationResponseAsync()` is not synchronous, so it might return the last notification response AFTER the response has come in and triggered the listener.

If you need to do something on each unique notification response, then your best bet is to make your hook dependent on the notification's `messageId` property. I'll do some testing to make sure that this property is being set correctly in all cases in the new FCMv1 flows, and test with your example above and make sure my suggestion works.

**Comment by itsramiel** [comment]
> I'll have a look at this example, to see why this behavior occurs. I've been thinking about this issue a bit more.
> 
> In general, I think it's probably expected that `getLastNotificationResponseAsync()` and `addNotificationResponseReceivedListener()` might have the same response, especially since `getLastNotificationResponseAsync()` is not synchronous, so it might return the last notification response AFTER the response has come in and triggered the listener.
> 
> If you need to do something on each unique notification response, then your best bet is to make your hook dependent on the notification's `messageId` property. I'll do some testing to make sure that this property is being set correctly in all cases in the new FCMv1 flows, and test with your example above and make sure my suggestion works.
> 
> If it works, I'll have a look at adding this uniqueness check in the `useLastNotificationResponse()` code itself, so that it works as you expect without having to change anything.

Perfect, thank you!

**Comment by itsramiel** [comment]
Hi @douglowder 

So I am now depending on the request identifier and that seems to work fine:

```ts
  const lastNotification = useLastNotificationResponse();

  useEffect(() => {
    if (lastNotification !== undefined && lastNotification !== null) {
      parseNotification(lastNotification.notification);
    }
  }, [lastNotification?.notification.request.identifier]);
```

>In general, I think it's probably expected that getLastNotificationResponseAsync() and >addNotificationResponseReceivedListener() might have the same response, especially since >getLastNotificationResponseAsync() is not synchronous, so it might return the last notification response AFTER the >response has come in and triggered the listener.

The problem is with incosistent behavior. 

`getLastNotificationResponseAsync` when used alone doesnt work and return nothing, but when a listener is attached with `addNotificationResponseReceivedListener` then it returns and also the listener returns the initial interacted notification. 

The other issue is the incosistency of `useLastNotificationResponse`. It returns the same notificaiton twice once from `getLastNotificationResponseAsync` and another from `addNotificationResponseReceivedListener` but **only on iOS**, on Android it it returns the initial notification once

**Comment by douglowder** [comment]
See latest commit in #30653, which adds the identifier check for `useLastNotificationResponse()`.

I'll have a look at why `getLastNotificationResponseAsync()` needs a listener.

**Comment by douglowder** [comment]
@itsramiel -  I've found the issue on iOS that causes `getLastNotificationResponseAsync()` to require a listener. Working on a fix.

Otherwise, things seem to be working as expected on both platforms.

**Comment by douglowder** [comment]
@itsramiel -- I've added a commit to #30653 that fixes the iOS issue with `getLastNotificationResponseAsync()`. It is still required to call `setNotificationHandler()` globally as in the beginning of our [documentation example code](https://docs.expo.dev/versions/latest/sdk/notifications/#usage). But now that API works correctly when no response listeners are defined.

**Comment by itsramiel** [comment]
Thank you @douglowder for your work. Looking forward to test this!

**Comment by douglowder** [comment]
Fix is now published in `expo-notifications@0.28.14`
